src/storage/mvcc/write.rs

```diff
@@ -102,21 +140,34 @@ impl Write {
         let write_type = WriteType::from_u8(b.read_u8()?).ok_or(Error::BadFormatWrite)?;
         let start_ts = number::decode_var_u64(&mut b)?;
         if b.is_empty() {
-            return Ok(Write::new(write_type, start_ts, None));
+            return Ok(Write::new_unprotected(write_type, start_ts, None));
         }
 
         let flag = b.read_u8()?;
-        assert_eq!(flag, SHORT_VALUE_PREFIX, "invalid flag [{}] in write", flag);
-
-        let len = b.read_u8()?;
-        if len as usize != b.len() {
-            panic!(
-                "short value len [{}] not equal to content len [{}]",
-                len,
-                b.len()
-            );
+        match flag {
+            SHORT_VALUE_PREFIX => {
+                let len = b.read_u8()?;
+                if len as usize != b.len() {
+                    panic!(
+                        "short value len [{}] not equal to content len [{}]",
+                        len,
+                        b.len()
+                    );
+                }
+                Ok(Write::new_unprotected(
+                    write_type,
+                    start_ts,
+                    Some(b.to_vec()),
+                ))
+            }
+            FLAG_PROTECTED => {
+                if write_type != WriteType::Rollback {
+                    panic!("only rollback record can be protected");
+                }
+                Ok(Write::new_rollback(start_ts, true))
+            }
+            _ => panic!("invalid flag [{}] in write", flag),
```

Comment:
I prefer to use `if-else` here since `protected` conflicts with short value. Now it seems they can exist simultaneously.

Reply:
Sorry, I don't get your point. Here we read the flag only once. I don't see in the code how they can exist simulaneously...

Reply:
Oh, my fault. Ignore it. :sweat_smile: